### ISSUE TITLE
Generalize SimLauncher resource locks

### ISSUE THREAD

**Issue by samreid** [phetsims/joist#621]
From https://github.com/phetsims/density/issues/13, we would like to generalize the SimLauncher resource locks.

> @samreid is going to take on unifying images, textures, phetioEngine, and brand with a time frame of about 2 weeks, and then pass it off to @jbphet to investigate the sound aspect.

**Comment by samreid** [comment]
I rewrote SimLauncher to use a list of locks and it is working well.  It allowed us to eliminate around 50 lines of very sketchy code in SimLauncher.  It also required changes to modulify.js and initialize-globals.js.  I tested by modulifying common code repos and several sims (including acid-base-solutions and density), and they worked well.  I tested adding a fake lock that is never unlocked and observed that the simulation never launched.

I think it would be safest if this code was reviewed before committing to master.  @zepumph would be a good reviewer since he is familiar with prior work in SimLauncher for ES6 support.  I think I'll put this code in branches to facilitate review.

To test, check out the branches.  Run grunt modulify on a sim and its dependencies (or on everything), test launching a simulation, read the code changes in SimLauncher, and modulify.  In particular, please check that the path logic `getDepth` in modulify.js works well on windows (it works well on mac, but I don't know what grunt.recurse is going to give for abspath on windows).  Branches incoming.

**Comment by zepumph** [comment]
I followed the above instructions, which were quite helpful. I couldn't run the sim until I merged chipper and joist masters into the branches. Then things worked well! I feel pretty confident that the merge conflicts were resolved correctly, but perhaps you want to take a look at the changes in modulify.js before merging to master.

* I ran `grunt modulify` on the following repos:

<details>

```
charges-and-fields
assert
axon
babel
brand
chipper
dot
joist
kite
phet-core
phetcommon
phetmarks
query-string-machine
scenery
scenery-phet
sherpa
sun
tambo
tandem
utterance-queue
```
</details>

* I added my own lock at the bottom of SimLauncher and saw that the sim didn't launch.
* I tested in phet brand
* I tested in phet-io brand standalone and in studio

**SimLauncher.js**

Elements that need some specific documentation:
* lockRemovedEmitter
* launchBegan
* launchComplete

* Why does `proceedIfReady` not assert that `launchComplete===false`? Is there a specific reason to be graceful there? If so please document.

* It is a bit confusing for `phetioEngine` to be defined at the bottom of the file. Perhaps make it at the top with "set at the bottom" doc?
* Since `phetioEngine` is set to null and only defined in phet-io brand. Should each usage in `launchSimulation()` include an assertion around it? Not sure how valuable that is since it would error out the line after.
* expanding on the above point. It is a bummer these dynamic imports need to go at the bottom. Any way around that? I don't think so but perhaps a bit more thought would be nice.
* We should use `Tandem.PHET_IO_ENABLED`  instead of the brand check. In https://github.com/phetsims/tandem/issues/153 we will make it run on brand, so it is best to be consistent.
* Does pendingLocks need to be a global forever? I would remove it.

 --------------------------------

**modulify.js**

* My `abspath` looks like `../joist/images/keyboard-icon.png` when logged in joist. Kinda weird! I wouldn't have called that an abspath personally.
* I feel like you should have the usage be a single function. What about `getPathDots` or `getExpandedDots( abspath)`. I don't see a reason to repeat `expandDots( getDepth( abspath ) )` each time.
* I know very little about image and (especially) mipmaps. That said the output of modulify looks really nice!

---------------------

* I would personally recommend creating a new issue to handle potentially using locks for sound. There is some nuance to discuss and I think it would be valuable to encapsulate that.

This implementation is really nice! Thanks for working on it. I definitely think that this is ready for master.

**Comment by samreid** [comment]
I mentioned I would work in this within the next couple of business days.  Bumping priority.

**Comment by samreid** [comment]
> My abspath looks like ../joist/images/keyboard-icon.png when logged in joist. Kinda weird! I wouldn't have called that an abspath personally.

That is the terminology used in grunt file.recurse, https://gruntjs.com/api/grunt.file#grunt.file.recurse. Should we do something else?

**Comment by samreid** [comment]
I addressed the review issues, aside from one which I commented on in a prior comment.  I merged chipper and joist to master and ran grunt modulify on everything.  Local aqua seems OK and I tested natural-selection in studio.  @zepumph can you please check the changes, mostly in SimLauncher.js and modulify.js?

**Comment by samreid** [comment]
I noticed balloons and static electricity qunit tests are failing CT:

```
balloons-and-static-electricity : top-level-unit-tests : require.js
https://bayes.colorado.edu/continuous-testing/ct-snapshots/1587854501239/balloons-and-static-electricity/balloons-and-static-electricity-tests.html
10 out of 11 tests passed. 1 failed.
WallDescriber: global failure failed:
Uncaught TypeError: this.callback is not a function
https://bayes.colorado.edu/continuous-testing/ct-snapshots/1587854501239/joist/js/SimLauncher.js:85


id: Bayes Chrome

balloons-and-static-electricity : top-level-unit-tests : require.js?ea
https://bayes.colorado.edu/continuous-testing/ct-snapshots/1587854501239/balloons-and-static-electricity/balloons-and-static-electricity-tests.html?ea
10 out of 11 tests passed. 1 failed.
WallDescriber: global failure failed:
Uncaught TypeError: this.callback is not a function
https://bayes.colorado.edu/continuous-testing/ct-snapshots/1587854501239/joist/js/SimLauncher.js:85


id: Bayes Chrome

balloons-and-static-electricity : top-level-unit-tests : require.js?ea
https://bayes.colorado.edu/continuous-testing/ct-snapshots/1587854501239/balloons-and-static-electricity/balloons-and-static-electricity-tests.html?ea
9 out of 10 tests passed. 1 failed.
WallDescriber: WallDescriber tests failed:
Uncaught TypeError: this.callback is not a function
https://bayes.colorado.edu/continuous-testing/ct-snapshots/1587854501239/joist/js/SimLauncher.js:85


id: Bayes Chrome

balloons-and-static-electricity : top-level-unit-tests : require.js?ea
https://bayes.colorado.edu/continuous-testing/ct-snapshots/1587854501239/balloons-and-static-electricity/balloons-and-static-electricity-tests.html?ea
9 out of 11 tests passed. 2 failed.
WallDescriber: global failure failed:
Uncaught TypeError: this.callback is not a function
https://bayes.colorado.edu/continuous-testing/ct-snapshots/1587854501239/joist/js/SimLauncher.js:85

WallDescriber: WallDescriber tests failed:
Died on test #1     at https://bayes.colorado.edu/continuous-testing/ct-snapshots/1587854501239/balloons-and-static-electricity/js/balloons-and-static-electricity/view/describers/WallDescriberTests.js:34:7: Assertion failed: boolean expected
Error: Assertion failed: boolean expected
    at window.assertions.assertFunction (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1587854501239/assert/js/assert.js:22:13)
    at supportsGestureDescription (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1587854501239/scenery-phet/js/accessibility/GrabDragInteraction.js:59:13)
    at new GrabDragInteraction (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1587854501239/scenery-phet/js/accessibility/GrabDragInteraction.js:220:38)
    at new BalloonNode (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1587854501239/balloons-and-static-electricity/js/balloons-and-static-electricity/view/BalloonNode.js:245:31)
    at new BASEView (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1587854501239/balloons-and-static-electricity/js/balloons-and-static-electricity/view/BASEView.js:83:28)
    at Object.<anonymous> (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1587854501239/balloons-and-static-electricity/js/balloons-and-static-electricity/view/describers/WallDescriberTests.js:38:16)


id: Bayes Chrome
```

**Comment by zepumph** [comment]
>That is the terminology used in grunt file.recurse, https://gruntjs.com/api/grunt.file#grunt.file.recurse. Should we do something else?

Nope!


I think that the closure you made called `initializeDynamicImports` is better than having them at the bottom, even though it is weird.


I made https://github.com/phetsims/joist/issues/630 for a refactor of SimLauncher into a singleton

I made https://github.com/phetsims/balloons-and-static-electricity/issues/466 for investigating the BASE issue. Even if it turns out to be general. I still think it is a good thing to be modular to this issue. 

Otherwise great work! Over to you to at least prioritize the other side issues. Feel free to assign me to them if you aren't going to work on them immediately.
